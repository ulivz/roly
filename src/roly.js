import fs from 'fs'
import * as rollup from 'rollup'
import switchy from 'switchy'
import chalk from 'chalk'
import getRollupOptions from './get-rollup-options'
import log from './log'
import {
  cwd,
  isString,
  isArray,
  splitStrByComma,
  handleRollupError,
  joinPath,
  isAbsolutePath
} from './utils'

function readInPkg(file) {
  try {
    const pkg = require(file)
    const roly = pkg.roly || {}
    if (pkg.name && !roly.filename) {
      const { name } = pkg
      roly.filename = name.startsWith('@')
        ? name.slice(name.lastIndexOf('/') + 1)
        : name
    }
    roly.pkg = pkg
    return roly
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      return {
        // Ensure pkg in options
        pkg: {}
      }
    }
    throw err
  }
}

function singleRoly(options = {}) {
  return new Promise(resolve => {
    options = Object.assign(
      {
        input: './src/index.js',
        format: ['cjs'],
        outDir: './dist',
        filename: 'index',
        compress: []
      },
      options
    )

    // for backward-compat
    if (options.entry) {
      options.input = options.entry
    }

    // for custom base dir
    const baseDir = options.baseDir
    if (baseDir) {
      if (!isAbsolutePath(options.input)) {
        options.input = joinPath(baseDir, options.input)
      }
      if (!isAbsolutePath(options.outDir)) {
        options.outDir = joinPath(baseDir, options.outDir)
      }
    }

    // Ensure format is an array
    let formats = options.format
    if (isString(formats)) {
      formats = formats === 'all'
        ? ['cjs', 'umd', 'es']
        : splitStrByComma(formats)
    } else if (!isArray(formats)) {
      throw new TypeError('Expect "format" to be a string or Array')
    }

    // Ensure compress is an array
    if (isString(options.compress)) {
      options.compress = splitStrByComma(options.compress)
    } else if (options.compress === true) {
      options.compress = ['cjs', 'umd', 'iife'] // Currently uglifyjs cannot compress es format
    } else if (!isArray(options.compress)) {
      throw new TypeError('Expect "compress" to be a string/true or Array')
    }

    options.compress = options.compress
      .filter(v => formats.includes(v))
      .map(v => `${v}Compress`)

    const allFormats = [...formats, ...options.compress]
    resolve(allFormats)
  }).then(allFormats => {
    return Promise.all(
      allFormats.map(format => {
        const rollupOptions = getRollupOptions(options, format)

        if (options.watch) {
          let init
          const watcher = rollup.watch(rollupOptions)
          return watcher.on('event', event => {
            switchy({
              START() {
                log(format, 'starting', chalk.blue)
                if (!init) {
                  init = true
                }
              },
              BUNDLE_START() {},
              BUNDLE_END() {
                log(format, 'bundled', chalk.green)
              },
              END() {},
              ERROR() {
                handleRollupError(event.error)
              },
              FATAL() {
                handleRollupError(event.error)
              },
              default() {
                console.error('unknown event', event)
              }
            })(event.code)
          })
        }

        return rollup.rollup(rollupOptions).then(bundle => {
          if (options.write === false)
            return bundle.generate(rollupOptions.output)
          return bundle.write(rollupOptions.output)
        })
      })
    ).then(result => {
      return result.reduce((res, next, i) => {
        res[allFormats[i]] = next
        return res
      }, {})
    })
  })
}

export default function roly(options = {}) {
  const baseDir = options.baseDir || ''
  const configFileName = options.config || 'roly.config.js'
  const configFilePath = cwd(joinPath(baseDir, configFileName))

  let rolyConfig = {}
  if (fs.existsSync(configFilePath)) {
    rolyConfig = require(configFilePath)
  }

  const pkgConfig = readInPkg(cwd(joinPath(baseDir, 'package.json')))

  if (Array.isArray(rolyConfig)) {
    return Promise.all(
      rolyConfig.map(rolyOptions => {
        return singleRoly(Object.assign(options, pkgConfig, rolyOptions))
      })
    )
  }
  return singleRoly(Object.assign(options, rolyConfig, pkgConfig))
}
