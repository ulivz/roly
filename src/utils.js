/* eslint-disable import/prefer-default-export */
import path from 'path'
import chalk from 'chalk'
import log from './log'

/**
 * Array type check
 */
export function isArray(arr) {
  return Array.isArray(arr)
}

/**
 * String type check
 */
export function isString(str) {
  return typeof str === 'string'
}

/**
 * Shorthand to get the abosulte path relative
 * to current process working directory
 */
export function cwd(...args) {
  return path.resolve(process.cwd(), ...args)
}

/**
 * String type check
 */
export function splitStrByComma(str) {
  return str.split(',').map(item => item.trim())
}

/**
 * path
 */
export const joinPath = path.join
export const isAbsolutePath = path.isAbsolute

/**
 * Handle rollup error
 */
export function handleRollupError(error) {
  console.log('123213')
  log(error.plugin || 'error', error.message, chalk.red)
  if (error.id) {
    console.log(chalk.dim(`Location: ${error.id}`))
  }
  console.error(error.snippet || error)
  console.log()
  process.exitCode = 1
}
