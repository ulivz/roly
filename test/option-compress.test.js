import {cwd} from './utils'
import roly from '../src/roly'

describe('compress', () => {
  it('true', async () => {
    const { umd, cjs, umdCompress, cjsCompress } = await roly({
      entry: cwd('fixtures/compress.js'),
      format: 'umd,cjs',
      compress: true,
      write: false
    })
    expect(umd.code).toMatchSnapshot()
    expect(cjs.code).toMatchSnapshot()
    expect(umdCompress.code).toMatchSnapshot()
    expect(cjsCompress.code).toMatchSnapshot()
  })

  it('string', async () => {
    const { umd, cjs, cjsCompress } = await roly({
      entry: cwd('fixtures/compress.js'),
      format: 'umd,cjs',
      compress: 'cjs',
      write: false
    })
    expect(umd.code).toMatchSnapshot()
    expect(cjs.code).toMatchSnapshot()
    expect(cjsCompress.code).toMatchSnapshot()
  })
})