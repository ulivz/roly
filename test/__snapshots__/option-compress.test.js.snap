// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compress string 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.roly = factory());
}(this, (function () { 'use strict';

var compress = function () { return 1; };

return compress;

})));
"
`;

exports[`compress string 2`] = `
"'use strict';

var compress = function () { return 1; };

module.exports = compress;
"
`;

exports[`compress string 3`] = `
"\\"use strict\\";var compress=function(){return 1};module.exports=compress;
"
`;

exports[`compress true 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.roly = factory());
}(this, (function () { 'use strict';

var compress = function () { return 1; };

return compress;

})));
"
`;

exports[`compress true 2`] = `
"'use strict';

var compress = function () { return 1; };

module.exports = compress;
"
`;

exports[`compress true 3`] = `
"!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):e.roly=n()}(this,function(){\\"use strict\\";return function(){return 1}});
"
`;

exports[`compress true 4`] = `
"\\"use strict\\";var compress=function(){return 1};module.exports=compress;
"
`;
